neo/io/blackrockio.py

```diff
@@ -1816,43 +1817,54 @@ def __read_analogsignal(
         description = \
             "AnalogSignal from channel: {0}, label: {1}, nsx: {2}".format(
                 channel_idx, labels[idx_ch], nsx_nb)
-
-        data_times = np.arange(
-            t_start.item(), t_stop.item(),
-            self.__nsx_basic_header[nsx_nb]['period']) * t_start.units
-        mask = (data_times >= n_start) & (data_times < n_stop)
-        data_times = data_times[mask].astype(float)
-
-        sig_ch = signal[dbl_idx][:, idx_ch][mask].astype(float)
-
-        # transform dig value to pysical value
-        sym_ana = (max_ana[idx_ch] == -min_ana[idx_ch])
-        sym_dig = (max_dig[idx_ch] == -min_dig[idx_ch])
-        if sym_ana and sym_dig:
-            sig_ch *= float(max_ana[idx_ch]) / float(max_dig[idx_ch])
+        
+        if lazy:
+            period = self.__nsx_basic_header[nsx_nb]['period']
+            length = int(((n_stop.rescale(nsx_time_unit)-n_start.rescale(nsx_time_unit))/\
+                            self.__nsx_basic_header[nsx_nb]['period']).magnitude)
```

Comment:
We believe that this expression does not necessarily reflect the shape of the resulting array.
Assume a period of 1 ms.

It works correctly for nstart=0.1 ms and nstop=5.0 ms. In this case the `length` computes above is 4. The statement in line 1834 covers samples at index 1, 2, 3, and 4. So the length is 4 (correct).

It works incorrectly for nstart=0.2 ms and nstop=5.1 ms. In this case the `length` computes above is still 4. The statement in line 1834 covers samples at index 1, 2, 3, 4, and 5. So the length is 5 (incorrect).

In our current version of the code, where we fixed the lazy loading in a somewhat similar fashion as you propose here, we used a very unelegant approach to circumvent this problem and get a consistency: We calculate `data_times` regardless of lazy or not, and use its length. This is not very efficient or nice, but maybe one can find a calculation that always works correctly.